scalafmt-core/shared/src/main/scala/org/scalafmt/Scalafmt.scala

```diff
@@ -62,48 +63,70 @@ object Scalafmt {
       else { // might throw for invalid conf
         val style = baseStyle.getConfigFor(filename)
-        val isSbt = filename.endsWith(".sc") || filename.endsWith(".sbt")
+        val isSbt = FileOps.isAmmonite(filename) || FileOps.isSbt(filename)
         if (isSbt) style.forSbt else style
       }
-    val runner = style.runner
-    Try {
-      if (code.matches("\\s*")) System.lineSeparator()
-      else {
-        val isWindows = containsWindowsLineEndings(code)
-        val unixCode = if (isWindows) {
-          code.replaceAll(WindowsLineEnding, UnixLineEnding)
-        } else {
-          code
-        }
-        val toParse = Rewrite(Input.VirtualFile(filename, unixCode), style)
-        val tree = runner.parse(toParse).get
-        val formatOps = new FormatOps(tree, style, filename)
-        runner.event(CreateFormatOps(formatOps))
-        val formatWriter = new FormatWriter(formatOps)
-        val partial = BestFirstSearch(formatOps, range, formatWriter)
-        val formattedString = formatWriter.mkString(partial.state)
-        val correctedFormattedString =
-          if (
-            (style.lineEndings == preserve && isWindows) ||
-            style.lineEndings == windows
-          ) {
-            formattedString.replaceAll(UnixLineEnding, WindowsLineEnding)
-          } else {
-            formattedString
-          }
-        if (partial.reachedEOF) {
-          correctedFormattedString
-        } else {
-          val pos = formatOps.tokens(partial.state.depth).left.pos
-          throw PreciseIncomplete(pos, correctedFormattedString)
-        }
-      }
-    } match {
+    val isWin = code.contains(WinLineEnding)
+    val unixCode =
+      if (isWin) code.replaceAll(WinLineEnding, UnixLineEnding) else code
+    doFormat(unixCode, style, range, filename) match {
       case Failure(e) => Formatted.Result(Formatted.Failure(e), style)
-      case Success(s) => Formatted.Result(Formatted.Success(s), style)
+      case Success(s) =>
+        val asWin = style.lineEndings == LineEndings.windows ||
+          (isWin && style.lineEndings == LineEndings.preserve)
+        val res = if (asWin) s.replaceAll(UnixLineEnding, WinLineEnding) else s
+        Formatted.Result(Formatted.Success(res), style)
     }
   }
 
+  private def doFormat(
+      code: String,
+      style: ScalafmtConfig,
+      range: Set[Range],
+      file: String
+  ): Try[String] =
+    if (!FileOps.isAmmonite(file)) doFormatOne(code, style, range, file)
+    else {
+      @tailrec
+      def iter(input: Seq[String], output: Seq[String]): Try[Seq[String]] =
+        if (input.isEmpty) Success(output)
+        else
+          doFormatOne(input.head, style, range, file) match {
+            case Failure(x) => Failure(x)
+            case Success(x) => iter(input.tail, output :+ x)
+          }
+      // see: https://ammonite.io/#Save/LoadSession
+      iter(code.split("(?:\\s*\\n@)+"), Seq.empty).map(_.mkString("\n@\n"))
```

Comment:
Just propose to generalize it as reusable function:
```scala
def traverse[A, B](xs: Seq[A])(f: A => Try[B]): Try[Seq[B]] = {
    @tailrec
    def iter(rest: Seq[A], res: Seq[B]): Try[Seq[B]] =
      if (rest.isEmpty) Success(res)
      else
        f(rest.head) match {
          case Success(x) => iter(rest.tail, res :+ x)
          case Failure(e) => Failure(e)
        }
    iter(xs, Seq.empty)
  }
```

and here

```scala
traverse(code.split("(?:\\s*\\n@)+"), code => doFormatOne(code, style, range, file)).map(_.mkString("\n@\n"))
```

Reply:
done. also, removed processing for ranges, because they wouldn't work after the split.